pkg/plugin/conntrack/_cprog/conntrack.c

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "vmlinux.h"
+#include "compiler.h"
+#include "bpf_helpers.h"
+#include "conntrack.h"
+
+
+/**
+ * The structure representing an ipv4 5-tuple key in the connection tracking map.
+ */
+struct ct_v4_key {
+    __u32 src_ip;
+    __u32 dst_ip;
+    __u16 src_port;
+    __u16 dst_port;
+    __u8 proto;
+};
+/**
+ * The structure representing a connection in the connection tracking map.
+ */
+struct ct_entry {
+    __u32 lifetime; // lifetime stores the time when the connection should be removed from the map.
+    /**
+     * last_report_*_dir stores the time when the last packet event was reported in the forward and reply direction respectively.
+     */
+    __u32 last_report_forward_dir;
+    __u32 last_report_reply_dir;
+    /**
+     * traffic_direction indicates the direction of the connection in relation to the host. 
+     * If the connection is initiated from within the host, the traffic_direction is egress. Otherwise, the traffic_direction is ingress.
+     */
+    enum ct_traffic_dir traffic_direction;
+    /**
+     * flags_seen_*_dir stores the flags seen in the forward and reply direction.
+     */
+    __u8  flags_seen_forward_dir;
+    __u8  flags_seen_reply_dir;
+    bool is_closing; // is_closing indicates if the connection is closing.
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __type(key, struct ct_v4_key);
+    __type(value, struct ct_entry);
+    __uint(max_entries, CT_MAP_SIZE);
+} retina_conntrack_map SEC(".maps");
+
+/**
+ * Returns the traffic direction based on the observation point.
+ * @arg observation_point The point in the network stack where the packet is observed.
+ */
+static __always_inline __u8 _ct_get_traffic_direction(enum obs_point observation_point) {
+    switch (observation_point) {
+        case FROM_ENDPOINT:
+        case TO_NETWORK:
+            return TRAFFIC_DIRECTION_EGRESS;
+        case TO_ENDPOINT:
+        case FROM_NETWORK:
+            return TRAFFIC_DIRECTION_INGRESS;
+        default:
+            return TRAFFIC_DIRECTION_UNKNOWN;
+    }
+}
+
+/**
+ * Create a new TCP connection.
+ * @arg key The key to be used to create the new connection.
+ * @arg flags The flags of the packet.
+ * @arg observation_point The point in the network stack where the packet is observed.
+ * @arg timeout The timeout for the connection.
+ */
+static __always_inline bool _ct_create_new_tcp_connection(struct ct_v4_key key, __u8 flags, enum obs_point observation_point, __u64 timeout) {
+    struct ct_entry new_value;
+    __builtin_memset(&new_value, 0, sizeof(struct ct_entry));
+    __u64 now = bpf_mono_now();
+    new_value.lifetime = now + timeout;
```

Comment:
you're storing 64bit values in a 32bit field - so `lifetime` does look like a unix timestamp. 
Add an explicit cast here to avoid a compiler warning if you're keeping the smaller field. I'd personally be ok with this - this gives Retina 80 years before overflow lol 
Let me ask the team offline.

Reply:
`bpf_ktime_get_boot_ns` officially returns u64, so i think I'll go ahead making everything time related u64

Reply:
Let's wait on others' feedback. ConnTrack introduces many timestamp values, and IMO it'd be wasteful to have several new half-empty 64-bit fields passed around.
The fact that the kernel uses 64-bit doesn't necessarily mean our internal representation needs to, as it can be safely cast down to 32-bit.

Reply:
And as above, regardless of what type is used in the end, you should be checking for overflow when using the sum operator